### ISSUE TITLE
RoaringBitmap.addOffset causing problems with serialization/deserialization

### ISSUE THREAD

**Issue by jeremycod** [RoaringBitmap/RoaringBitmap#260]
Hi,

Based on Issue #251 I've created a utility method that aggregates two bitmaps
 
```
  override def concatenateBitmaps(index: Int, resultBitSets: Array[BitSet], numRows: Int, offset: Int): Unit = {
    val shifted = RoaringBitmap.addOffset(underlying.toRoaringBitmap, offset)
    val underlyingBitSet = resultBitSets(index).underlying.asInstanceOf[MutableRoaringBitmap].toRoaringBitmap
    underlyingBitSet.or(shifted)
    resultBitSets(index) = new RoaringBitSet(underlyingBitSet.toMutableRoaringBitmap, numRows)
  }
```
 
Results of this operation are serialized and deserialized later again. Sometimes, it hits EOFException and I can't deserialize bitmap from the file. I was able to reproduce the problem in a simple unit test that I'm sending in attachment. 
It seems that the problem is in this part:

`val shifted = RoaringBitmap.addOffset(underlying.toRoaringBitmap, offset)`

With test data I've provided and offset 5950 shifted value that I get as a result has BitMapContainer with actual data and additional ArrayContainer which is empty. Even debugger shows exception when calling RoaringBitmap.toString(), but main issue is when I serialize and try to deserialize such bitmap, it always produces exception:

```
[info] RoaringBitSetsSerializationProblemSpec:
[info] - should serialize and deserialize the same roaring bitmap *** FAILED ***
[info]   java.io.EOFException:
[info]   at java.io.DataInputStream.readFully(DataInputStream.java:197)
[info]   at java.io.DataInputStream.readLong(DataInputStream.java:416)
[info]   at java.io.ObjectInputStream$BlockDataInputStream.readLong(ObjectInputStream.java:3163)
[info]   at java.io.ObjectInputStream.readLong(ObjectInputStream.java:1034)
[info]   at org.roaringbitmap.buffer.MutableRoaringArray.deserialize(MutableRoaringArray.java:295)
[info]   at org.roaringbitmap.buffer.MutableRoaringArray.readExternal(MutableRoaringArray.java:499)
[info]   at org.roaringbitmap.buffer.MutableRoaringBitmap.readExternal(MutableRoaringBitmap.java:1269)
[info]   at visier.platform.data.bitset.RoaringBitSet.deserialize(BitSet.scala:244)
[info]   at visier.engine.BitSetsSerializationProblemSpec$$anonfun$1$$anonfun$apply$mcV$sp$3.apply(BitSetsSerializationProblemSpec.scala:60)
[info]   at visier.engine.BitSetsSerializationProblemSpec$$anonfun$1$$anonfun$apply$mcV$sp$3.apply(BitSetsSerializationProblemSpec.scala:59)
```
I'm using RoaringBitmap 0.7.12

Any ideas how to solve this problem?

![screen shot 2018-06-21 at 1 45 20 pm](https://user-images.githubusercontent.com/5204745/41744708-7913932c-7559-11e8-9aa4-7740fd162991.png)

[roaring.txt](https://github.com/RoaringBitmap/RoaringBitmap/files/2125250/roaring.txt)

[BitSetsSerializationProblemSpec.txt](https://github.com/RoaringBitmap/RoaringBitmap/files/2125263/BitSetsSerializationProblemSpec.txt)

**Comment by lemire** [comment]
Given that you have provided code, it is likely a bug. I am not sure how we might end up with an empty container. That is certainly a bug.

So we need to fix this.

(I am traveling, if anyone sees this and gets to it before I do, please investigate.)

**Comment by richardstartin** [comment]
I looked at this briefly, added an assert in `RoaringArray.append` and wrote a couple of fuzz tests to try and trigger it. It looks like there are several other bugs exposed by or introduced by this new function:

- ensuring capacity during `RunContainer.append`
- addOffset can return a `BitmapContainer` in a lazy state and later calls `cardinalityInRange` which asserts the bitmap is not lazy

**Comment by richardstartin** [comment]
I actually could not reproduce the issue with the data uploaded here, but found a few more bugs. See https://github.com/RoaringBitmap/RoaringBitmap/pull/261/, and let me know if the test case loading the data from the file adequately models the reported bug. I don't have time to track down the causes of the "awkward offsets" test failures.

**Comment by jeremycod** [comment]
I see this is merged, so I guess it's solved. When will this be in release?

**Comment by richardstartin** [comment]
@jeremycod I'm not sure. I couldn't reproduce your issue with the data you supplied so can't say if this will actually fix your problem... but it does fix a few other issues that would affect you eventually. Maybe you could build from source locally and check with asserts enabled?

**Comment by richardstartin** [comment]
I added test cases which serialise and deserialise via a `byte[]` for both [RoaringBitmap ](https://github.com/RoaringBitmap/RoaringBitmap/blob/master/roaringbitmap/src/test/java/org/roaringbitmap/TestConcatenation.java#L100)and [MutableRoaringBitmap](https://github.com/RoaringBitmap/RoaringBitmap/blob/master/roaringbitmap/src/test/java/org/roaringbitmap/TestConcatenation.java#L110), parametrised with your data and offset [here](https://github.com/RoaringBitmap/RoaringBitmap/blob/master/roaringbitmap/src/test/java/org/roaringbitmap/TestConcatenation.java#L28). Is there something missing?

As @lemire states, if there is an empty container at any point, it's definitely a bug, but I don't think the combination of (your data, offset=5950, use `addOffset`) is enough to reproduce it. It would be very helpful if you could submit a PR with a reproducing test case, in which case I would be more than happy to help with a fix.

**Comment by richardstartin** [comment]
I just reproduced the issue on master prior to merging with this test:

```java
  @Test
  public void test260() {
    IntStream indices = read("src/test/resources/testdata/testIssue260.txt");
    MutableRoaringBitmap bitmap = new MutableRoaringBitmap();
    indices.forEach(bitmap::add);
    int offset = 5950;
    RoaringBitmap shifted = RoaringBitmap.addOffset(bitmap.toRoaringBitmap(), offset);
    Arrays.stream(shifted.highLowContainer.values).forEach(c -> assertFalse(c.isEmpty()));
  }


  private static IntStream read(String classPathResource) {
    try {
      return Arrays.stream(Files.readFirstLine(new File(classPathResource), Charset.forName("UTF-8")).split(","))
              .mapToInt(Integer::parseInt);
    } catch (IOException e) {
      throw new RuntimeException(e);
    }
  }
```

It's fixed by #261 anyway. @lemire it might be worth doing a release.

**Comment by lemire** [comment]
Preparing a release.

**Comment by lemire** [comment]
(Fantastic work @richardstartin)

**Comment by jeremycod** [comment]
Awesome. Thanks a lot for solving it so quickly.

**Comment by lemire** [comment]
Fixed as of RoaringBitmap-0.7.14 

Please reopen if you cannot verify.
